### PR TITLE
[USH-1137] Data view: Stabilize UI & retain desktop, tablet or mobile look on resize

### PR DESCRIPTION
**Pull request by Ifycode**
**Expected Result & Fixes (A - F below):**

> **_Please note while reviewing or testing:_** Modes have changed. The names of the modes are no longer tied to the look of the data view pages. Instead, they reflect what we expect the user to be able to do, interact with, access or see on those pages, which favours long-term or long-run use. If we decide in the future that the pages should look differently from what we currently have (as long as what we expect users to access on that page has not changed), we won't need to bother about changing the mode names. The naming convention also helps to understand the "make up" or "building blocks" of the modes, as well as the purpose of the modes to help identify single page or a set of pages.
> 
> New names of modes and what they represent:
> 
> **_ID mode:_**
> - page URLs/UIs that involve the user identifying/accessing a single unique post by its ID to **_view_** full details or to **_edit_** the post details.
> - page URLs/UIs for inability to access a single unique post by ID, because its either not/no-longer existing or it exists but user does not have permissions to access such post.
>
> **_PREVIEW mode:_** 
> - page URL/UI that involve user viewing multiple posts in their preview form, when user is not (yet) trying to access the posts by their ID.

**A. Browser resize fix (1 of 3):** 

Predictable & consistent Browser URL across different devices, when you resize browser from desktop - to tablet - to mobile, and vice versa. Check for the following scenarios:

- **_When both pagination and collections are not involved:_**
    - For post cards preview-only page (**_a.k.a PREVIEW mode_**), but without `mode=PREVIEW` in browser URL: 
        - Go to `/feed`, then resize browser. See that the same URL is what you will find in browser URL through out the resize.
    - Still on `/feed`, but with `mode=` present in browser URL:
        - Go to `/feed?mode=PREVIEW`, then resize browser. See that the same URL is what you will find in browser URL through out the resize.
        - For an **_existing post that user has permission to access by ID_** (**_a.k.a ID mode_**):
            - Go to `/feed/post-id-goes-here/view?mode=ID` to **_view_** any post of your choice, then resize browser. See that the same URL is what you will find in browser URL through out the resize.
            - Go to `/feed/post-id-goes-here/edit?mode=ID` to **_edit_** any post of your choice, then resize browser. See that the same URL is what you will find in browser URL through out the resize.
- **_When pagination is involved, but collections is not involved:_**
    - For post cards preview-only page (**_a.k.a PREVIEW mode_**), but without `mode=PREVIEW` in browser URL: 
        - Go to `/feed?page=page-number-goes-here`, then resize browser. See that the same URL is what you will find in browser URL through out the resize.
    - Still on `/feed`, but with `mode=` present in browser URL:
        - Go to `/feed?page=page-number-goes-here&mode=PREVIEW`, then resize browser. See that the same URL is what you will find in browser URL through out the resize.
        - For **_an existing post that user has permission to access by ID_** (**_a.k.a ID mode_**):
            - Go to `/feed/post-id-goes-here/view?page=page-number-goes-here&mode=ID` to **_view_** any post of your choice, then resize browser. See that the same URL is what you will find in browser URL through out the resize.
            - Go to `/feed/post-id-goes-here/edit?page=page-number-goes-here&mode=ID` to **_edit_** any post of your choice, then resize browser. See that the same URL is what you will find in browser URL through out the resize.
- **_When collections is involved, but pagination is not involved:_**
    - For post cards preview-only page (**_a.k.a PREVIEW mode_**), but without `mode=PREVIEW` in browser URL: 
        - Go to `/feed/collection/collection-id-goes-here`, then resize browser. See that the same URL is what you will find in browser URL through out the resize.
    - Still on `/feed/collection`, but with `mode=` present in browser URL:
        - Go to `/feed/collection/collection-id-goes-here?mode=PREVIEW`, then resize browser. See that the same URL is what you will find in browser URL through out the resize.
        - For an **_existing post (that has been added to a collection and) that user has permission to access by ID_** (**_a.k.a ID mode_**):
            - Go to `/feed/collection/collection-id-goes-here/post-id-goes-here/view?mode=ID` to **_view_** any collection post of your choice, then resize browser. See that the same URL is what you will find in browser URL through out the resize.
            - Go to `/feed/collection/collection-id-goes-here/post-id-goes-here/edit?mode=ID` to **_edit_** any collection post of your choice, then resize browser. See that the same URL is what you will find in browser URL through out the resize.
- **_When both collections and pagination are involved:_**
    - For post cards preview-only page (**_a.k.a PREVIEW mode_**), but without `mode=PREVIEW` in browser URL: 
        - Go to `/feed/collection/collection-id-goes-here?page=page-number-goes-here`, then resize browser. See that the same URL is what you will find in browser URL through out the resize.
    - Still on `/feed/collection`, but with `mode=` present in browser URL:
        - Go to `/feed/collection/collection-id-goes-here?page=page-number-goes-here&mode=PREVIEW`, then resize browser. See that the same URL is what you will find in browser URL through out the resize.
        - For **_an existing post (that has been added to a collection and) that user has permission to access by ID_** (**_a.k.a ID mode_**):
            - Go to `/feed/collection/collection-id-goes-here/post-id-goes-here/view?page=page-number-goes-here&mode=ID` to **_view_** any collection post of your choice, then resize browser. See that the same URL is what you will find in browser URL through out the resize.
            - Go to `/feed/collection/collection-id-goes-here/post-id-goes-here/edit?page=page-number-goes-here&mode=ID` to **_edit_** any collection post of your choice, then resize browser. See that the same URL is what you will find in browser URL through out the resize.

**B. Browser resize fix (2 of 3):** 

Data view's **_ID mode_** pages UI changes consistently on browser resize - Desktop/larger devices have post **_details_** or post **_edit_** component on the right of the page, smaller devices only use modal popup for showing the post **_details_** or post **_edit_** component.

Following the idea from url fixes above, Check that UI cooperates on resize as described:
- When both pagination and collections are not involved
- When pagination is involved, but collections is not involved
- When collections is involved, but pagination is not involved
- When both collections and pagination are involved

**C. Browser resize fix (3 of 3):**

Resize fix also implemented for new routes added based on the inability to access a single unique post by ID, because its either not/no-longer existing or it exists but user does not have permissions to access such post.

- When user is trying to **_view_** a post that doesn't exist i.e. **_not-found_**: Go to url containing a non-existing post ID `/feed/non-exisiting-post-id-goes-here/view?mode=ID`, you will be redirected to `/feed/non-exisiting-post-id-goes-here/not-found?mode=ID`. Upon redirect, resize browser and check that:
    - The same **_not-found_** URL is what you will find in browser URL through out the resize.
    - UI cooperates on resize.
- When user is **_not-allowed_** to **_edit_** a post is trying to access the `/edit` route: As a Guest user or user that does not have permission to manage/edit all posts (and may or may not have manage/edit their own posts) - Go to edit url of an existing post, that you have access to view, but not access to edit `/feed/exisiting-post-id-I-can-view-but-dont-have-permission-to-edit-goes-here/edit?mode=ID`, you will be redirected to `/feed/exisiting-post-id-I-can-view-but-dont-have-permission-to-edit-goes-here/not-allowed?mode=ID`. Upon redirect, resize browser and check that:
    - The same **_not-allowed_** URL is what you will find in browser URL through out the resize.
    - UI cooperates on resize.
- Following the ideas written earlier, you can check that both **_browser URL_** and **_UI_** cooperates for **_not-allowed_** and **_not-found_** too on resize as described:
    - When both pagination and collections are not involved
    - When pagination is involved, but collections is not involved
    - When collections is involved, but pagination is not involved
    - When both collections and pagination are involved

**D. Resize fix (extra notes):**
- You can also check resize for when mode params come before page params in the browser URL like so:
    - Preview mode: `?mode=PREVIEW&page=page-number-goes-here`
    - ID mode: `?mode=ID&page=page-number-goes-here`
- How to access the browser URL routes listed earlier, to check for the UI and Browser URL fixes:
    - Through **_page load/reload_** (for all browser URL routes)
    - In the case of  `/feed`, also through clicking the **_sidebar link_** for data view
    - Through clicking on the **_pagination buttons_** in PREVIEW mode - this introduces/adds `page=page-number` in the browser URL
    - In the case of **_ID mode routes_**, also:
        - Through clicking on a post card **_body_** from PREVIEW mode - this introduces/adds `mode=ID` in the browser URL
        - Through clicking on the **_edit icon_** on a post card from PREVIEW mode - this also introduces/adds `mode=ID` in the browser URL
        - Through clicking any of the **_"switch mode" buttons_** (available on desktop only) - this introduces/adds `mode=PREVIEW` or `mode=ID` in the browser URL depending on the mode you are entering
- You can interact with filters too, and then resize afterwards.
   [**_note:_** a different issue is to take care of some blockage on resize to smaller devices - ​​[USH-1413](https://linear.app/ushahidi/issue/USH-1413/cant-interact-with-both-filters-and-post-detailsedit-modals-in-smaller)]

**E. Plus Fix 🎁🥳:**

This Fix automatically takes care of these too, to "match" with refactoring done while working on the browser resize issues:

- **_Regardless of device used:_**
    - **_New redirects for post by ID:_**
        - **New **_not-found_** route:**
            - Initial redirect or page load/reload:
                - Redirect to new **_not-found_** route happens, when user is trying **_view_** a post that doesn't exist - See ticket for aspects still in discussion and to be fixed later on - [USH-1384](https://linear.app/ushahidi/issue/USH-1384/data-views-view-route-separating-ui-for-not-found-in-the-deployment)
                - Loading/reloading page while still on the **_not-found_** route you have been redirected to retains not-found browser URL and UI content
                - **_Pagination_** with **_not-found_** routes: While still on the **_not-found_** route you have been redirected to, adding a specific page in the browser URL (and re-loading the page) retains not-found browser URL and UI content, but will also show posts and posts results from the page specified
            - Redirects **_back to other routes_**:
                - Using the post ID for post that exists and that a user is allowed to view in the not-found url, will redirect user to the post's `/view` route accordingly
                - Using the post ID for post that user is not allowed to **_edit_** in the not-allowed url, will redirect user to the **_not-allowed_** route accordingly
        - New **_not-allowed_** route:
            - Initial redirect or page load/reload:
                - Redirect to new **_not-allowed_** route happens, when user who is **_not-allowed_** to **_edit_** a post is trying to access the post's `/edit` route
                - Loading/reloading page while still on the **_not-allowed_** route you have been redirected to retains not-allowed browser URL and UI content
                - **_Pagination_** with **_not-allowed_** routes: While still on the **_not-allowed_** route you have been redirected to, adding a specific page in the browser URL (and re-loading the page) retains not-allowed browser URL and UI content, but will also show posts and posts results from the page specified.
            - Redirects **_back to other routes_**:
                - Using the post ID for post that exists and that a user is allowed to view in the not-allowed url, will redirect user to the post's `/view` route accordingly
                - Using the post ID for post non-existing post in the not-allowed url, will redirect user to the **_not-found_** route accordingly
    - **_Mode-related changes:_**
        - **_Mode name change:_** See explanation done in **_"Please note while reviewing or testing"_** notes earlier.
        - **_Browser redirects if not one of the 2 modes:_** If user supplies another mode in the URL other than the mode options we allow e.g. `mode=BATCAVE`, it will redirect back to the current mode that the user is in.
- **_Smaller devices only:_**
    - Closing post details/edit modal popup redirects back to PREVIEW mode (browser URL and UI).
- **_Desktop/larger devices only:_**
    - Quicker access and clean UI switch to ID mode routes (`/view` or `/edit`) from Preview mode (and vice versa) i.e. only the post details/edit component load, not the entire posts loading along - regardless of whether pagination or collection is involved or not. To see this change, access **_ID mode routes_**:
        - Through clicking on a post card **_body_** from PREVIEW mode
        - Through clicking on the **_edit icon_** on a post card from PREVIEW mode
        - Through clicking any of the **_"switch mode" buttons_**
    - Loader(s) that take up the entire posts content area only now show up in this cases:
        - On page load/reload
        - Upon interaction with filters or clearing of filters

**F. Engineering-specific notes:** 

Major code changes (based on the fix that permitted UI to work well on browser resize):
- **_Code perception and interpretation:_** Previously, code was perceived based on the look of the UI when accessing modes, and which you will find is also reflected in the browser URL (this was particularly noticeable particularly for smaller devices, technically there was no ID mode for smaller device). The new fixes help see/perceive/interpret code based on if a user wants just previews of posts (only) or they just want to access a single unique post by its ID - **_browser URL_** now match with what user can access through a mode's **_UI_**, regardless of the device size.
- **_The router:_** Router now controls everything router-related - i.e. as long as the user activity triggers the main data view route `/feed` or any of the child routes `/view`  `/edit`, **_regardless of if the routing is triggered by a button/card click or page load/reload_** (With the exception of the "Load more" button and filters [apart from "Clear all filters" button] since these do not trigger a route change):
    - **_Router sets the initial states_** for variables like `this.isLoading`, `this.mode`, and other variables/properties we want to use to change the UI - i.e. How we set states are set initially in just one place (in the constructor and at route level, instead of having to always set it everywhere we needed such initial state values - as was the case before).
    - **_Router uses a check for post ID to determine the data view mode_** (i.e. `this.mode`) that user is in (more reliable than previous method of manually setting & getting mode from queryParams - anytime there was some confusion in the code, UI or browser url [and that affected browser resize], this was the main culprit 🙃 Data view's structure is too complex to be setting anything like that manually🙃). This also influenced the adoption of the mode name "ID mode". It also makes it easier to redirect users to the correct mode for the page they are on, if they try out using modes not in the option for them in the browser url.
    - **_Router employs a guard_** `RedirectByPostIdGuard`, to handle redirects to (ID mode) **_child routes_** in data view (including the newly added **_not-found_** and **_not-allowed_** routes). More details in **"E. Plus Fix 🎁🥳"** section about workings and benefits of using the redirect guard.
- **_API response:_**
    - Letting API response determine when e.g. `this.isLoading` state returns false.
    - Using `finalize()` to access API response, to prevent pass through when error occurs (which will help to better project error to users in a very readable form, when we are ready to do so).

### CODE REVIEW

**Comment by Ifycode** [comment]
PR finally ready for review @Angamanga @ushahidlee 🥳

File: apps/web-mzima-client/src/app/feed/feed.component.ts [hunk 1]
```diff
@@ -282,134 +425,279 @@ export class FeedComponent extends MainViewComponent implements OnInit {
...
+        }, 70);
+      } else {
+        postVisibility(post, '');
+      }
+    });
+
+    /* -------------------------------------------
+        Never (ever) forget this "guy" when you need 
+        styles to adjust for masonry library
+      --------------------------------------------*/
```
Comment (Angamanga) [inline apps/web-mzima-client/src/app/feed/feed.component.ts:472]: 😆

File: apps/web-mzima-client/src/app/feed/feed.component.ts [hunk 2]
```diff
@@ -25,17 +26,21 @@ import {
 import _ from 'lodash';
 
 enum FeedMode {
-  Tiles = 'TILES',
-  Post = 'POST',
+  Preview = 'PREVIEW',
+  Id = 'ID',
 }
```
Comment (Angamanga) [inline apps/web-mzima-client/src/app/feed/feed.component.ts:31]: Is this "just" to make naming more general?

Comment (Angamanga) [reply apps/web-mzima-client/src/app/feed/feed.component.ts:31]: What is the difference between /feed and /feed?mode=PREVIEW? Are they always directing to the tiles? Is PREVIEW really needed if it is the same thing?

**Review by Angamanga — COMMENTED** [review]
Added some questions in the code, but mostly works really well. I noticed a tiny malfunction on the not-allowed/not-found routes: The modal when resizing is empty, probably a "small" thing to fix 😀 
Screenshot 2024-06-24 at 06 47 02

File: apps/web-mzima-client/src/app/feed/feed.component.ts (hunk 1 above)
Comment (Ifycode) [reply apps/web-mzima-client/src/app/feed/feed.component.ts:472]: 😆

File: apps/web-mzima-client/src/app/feed/feed.component.ts (hunk 2 above)
Comment (Ifycode) [reply apps/web-mzima-client/src/app/feed/feed.component.ts:31]: Replying to https://github.com/ushahidi/platform-client-mzima/pull/1053#discussion_r1642740168 - The Preview and ID terms basically does 2 things:
- Reflect the code directly on the UI/browser URL and also the other way round. For example, as long as there is a post ID number in the browserURL, you know you are in "ID mode" (regardless of what the UI looks like) - this is from the user perspective. From the code perspective, the router also uses the check for ID to know whether you want to access ID mode pages or not.
- The "regardless of what the UI looks like" part - is just based on what the user can access from the UI without any focus on what the UI looks like. ID mode - the user wants to access the pages that use post ID in the browser URL. Preview mode - The user wants to access the page that don't use post ID in the browser URL, and on this page you can only find the preview of posts (which can be any any form - card, table etc.)

Comment (Ifycode) [reply apps/web-mzima-client/src/app/feed/feed.component.ts:31]: Replying to https://github.com/ushahidi/platform-client-mzima/pull/1053#discussion_r1642741770 - It's just a movement to `/feed` as directed by the router, with mode params attached/added - similar to how page params is added when there's pagination.
If it's needed - it depends on how you see it. But what I can say is that it was really helpful for debugging the issues with browser URL.

Comment (Ifycode) [reply apps/web-mzima-client/src/app/feed/feed.component.ts:31]: I'm not able to see your screenshot here @Angamanga - https://github.com/ushahidi/platform-client-mzima/pull/1053#pullrequestreview-2122779167
Me and my github images/videos problems...

Comment (Angamanga) [reply apps/web-mzima-client/src/app/feed/feed.component.ts:31]: Comment for any future reader: We discussed this https://github.com/ushahidi/platform-client-mzima/pull/1053#pullrequestreview-2122779167 in Slack and there is an issue for it here:
https://linear.app/ushahidi/issue/USH-1413/cant-interact-with-both-filters-and-post-detailsedit-modals-in-smaller

**Review by Angamanga — APPROVED** [review]

**Comment by Angamanga** [comment]
@Ifycode @shakirandagire The structure of the data-view has changed a bit and I see there are e2e-tests failing. Could you go through the changes together and see if some tests need an update?

**Comment by Ifycode** [comment]
Updating branch and merging, as discussed on the sprint call.

**Pull request merged**